src/main/scala/tapasco/filemgmt/ProgressTrackingFileWatcher.scala

```diff
@@ -0,0 +1,115 @@
+
+import de.tu_darmstadt.cs.esa.tapasco.Logging._
+import de.tu_darmstadt.cs.esa.tapasco.util.Listener
+import MultiFileWatcher._
+import Events._
+import ProgressTrackingFileWatcher._
+
+import scala.concurrent.duration.Duration
+
+/**
+  * A [[MultiFileWatcher]] which tracks the overall logfile.
+  * Using a FSM, the current state of the Compose Progression is Determined and Logged
+  * to give the user an overview how far the compose-Task has progressed.
+  * @param _logger Optional logger instance to use.
+  * @param pollInterval Polling interval in ms (default: [[MultiFileWatcher.POLL_INTERVAL]]).
+  */
+class ProgressTrackingFileWatcher(_logger: Option[Logger] = None, pollInterval: Int = POLL_INTERVAL)
+  extends MultiFileWatcher(POLL_INTERVAL) {
+  private[this] final val logger = _logger getOrElse de.tu_darmstadt.cs.esa.tapasco.Logging.logger(getClass)
+
+  var currentState: Int = -1
+  val start: Long = System.currentTimeMillis()
+  var stageStart: Long = System.currentTimeMillis()
+  var lastLog: Long = 0
+
+  private lazy val listener = new Listener[Event] {
+    def update(e: MultiFileWatcher.Event): Unit = e match {
+      case LinesAdded(src, ls) => {
+        ls foreach {l =>
+          checkState(l)
+        }
+      }
+    }
+  }
+
+  /**
+    * Checks whether a transition has occured depending on the current line of the log.
+    * If so, the corresponding new state is logged.
+    * Additionally, the finished State is also logged with its runtime.
+    * @param string parsed String
+    */
+  private def checkState(string: String): Unit = {
+    val old = currentState
+    if(currentState == -1) {
+      currentState = currentState + 1
+    }
+    for(i <- progressionStringsInfo.indices) {
+      val progressionString = progressionStringsInfo(i)._1
+      if(i == -1 || (currentState == i - 1 && string.contains(progressionString))){
+        currentState = currentState + 1
+      }
+    }
+    if(old != currentState) {
+      if(old != -1) {
+        val oldProgressionString = progressionStringsInfo(old)._2
+        logger.info("Finished %s after %s (Total Elapsed: %s)".format(oldProgressionString, timeString(stageStart), timeString(start)))
+      }
+      val progressionString = progressionStringsInfo(currentState)._2
+      logger.info("Started %s (Total Elapsed: %s)".format(progressionString, timeString(start)))
+      stageStart = System.currentTimeMillis()
+
+      //This ensures that this watcher will close all opened files if he reaches the final state.
+      if(currentState == progressionStringsInfo.length) {
+        closeAll()
+      }
+
+
+    }
+  }
+
+  /**
+    * Calculates the time since a certain TimeStamp and transforms it into a hh:mm:ss-Formatted String.
+    * @param since timestamp
+    * @return TimeString
+    */
+  private def timeString(since: Long): String = {
+    val now = System.currentTimeMillis()
+    val dur = Duration(now-since, "millis")
+    val s = dur.toSeconds.toInt
```

Comment:
Why do you first get the duration in milliseconds if you anyway convert it to seconds immediately after that? You could get the number of seconds with `"seconds"` as parameter to `Duration.apply`.

Reply:
This is now more consistent. `Duration.apply` takes a Long with the duration and a string indicating the unit. The `millis` indicates that the input is in ms.
The `Duration` object is then used to transform into the other units (hours, minutes and seconds).